lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,285 @@
+      end
+
+      def add_amex(post, options)
+        if options[:amex].present?
+          post[:amex] = {}
+          post[:amex][:propertyCode] = options[:amex][:property_code]
+        end
+      end
+
+      def add_auto(post, options)
+        if options[:auto].present?
+          post[:auto] = {}
+          post[:auto][:estimatedDays] = options[:auto][:estimated_days]
+        end
+      end
+
+      def add_card_on_file(post, options)
+        if options[:stored_credential].present?
+          post[:cardOnFile] = {}
+          post[:cardOnFile][:indicator] = options[:stored_credential][:indicator] if options[:stored_credential][:indicator]
+          if usage_indicator = options[:stored_credential][:usage_indicator] == 'initial' ? '01' : '02'
+            post[:cardOnFile][:usageIndicator] = usage_indicator
+          end
+          if scheduled_indicator = options[:stored_credential][:scheduled_indicator] == 'scheduled' ? '' : '02'
+            post[:cardOnFile][:scheduledIndicator] = scheduled_indicator
+          end
+          post[:cardOnFile][:transactionId] = options[:stored_credential][:transaction_id] if options[:stored_credential][:transaction_id]
+        end
+      end
+
+      def commit(action, parameters, option)
+        url_postfix = action == 'accesstoken' ? 'credentials' : 'transactions'
+        url = (test? ? "#{test_url}#{url_postfix}/#{action}" : "#{live_url}#{url_postfix}/#{action}")
+        if action == 'invoice'
+          response = parse(ssl_request(:delete, url, parameters.to_json, request_headers(option)))
+        else
+          response = parse(ssl_post(url, parameters.to_json, request_headers(option)))
+        end
+
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response, action),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def handle_response(response)
+        case response.code.to_i
+        when 200...300, 400, 401
+          response.body
+        else
+          raise ResponseError.new(response)
+        end
+      end
+
+      def parse(body)
+        return {} if body == ''
+
+        JSON.parse(body)
+      end
+
+      def message_from(response)
+        success_from(response) ? 'Transaction successful' : error(response)['longText']
+      end
+
+      def error_code_from(response)
+        return unless success_from(response)
+
+        STANDARD_ERROR_CODE_MAPPING[response['primaryCode']]
+      end
+
+      def authorization_from(response, action)
+        response.dig('result', 0, 'card', 'token', 'value')
+      end
+
+      def request_headers(options)
+        headers = {
+          'Content-Type' => 'application/x-www-form-urlencoded'
+        }
+        headers['CompanyName'] = options[:company_name]
+        headers['AccessToken'] = options[:access_token]
```

Comment:
https://github.com/activemerchant/active_merchant/pull/4415/files#r897876783

After updating you can get access toke from @options[:access_token]